proxy/src/transparency.rs

```diff
@@ -0,0 +1,688 @@
+        let auth = Authority::from_shared(bytes)
+            .expect("socket address is valid authority");
+        set_authority(req.uri_mut(), auth);
+
+        return Ok(());
+    }
+
+    Err(())
+}
+
+fn set_authority(uri: &mut http::Uri, auth: Authority) {
+    let mut parts = Parts::from(mem::replace(uri, Uri::default()));
+    parts.scheme = Some(Scheme::HTTP);
+    parts.authority = Some(auth);
+
+    let new = Uri::from_parts(parts)
+        .expect("absolute uri");
+
+    *uri = new;
+}
+
+fn strip_connection_headers(headers: &mut http::HeaderMap) {
+    let conn_val = if let Some(val) = headers.remove(http::header::CONNECTION) {
+        val
+    } else {
+        return
+    };
+
+    let conn_header = if let Ok(s) = conn_val.to_str() {
+        s
+    } else {
+        return
+    };
+
+    // A `Connection` header may have a comma-separated list of
+    // names of other headers that are meant for only this specific connection.
+    //
+    // Iterate these names and remove them as headers.
+    for name in conn_header.split(',') {
+        let name = name.trim();
+        headers.remove(name);
+    }
+}
+
+
+// ==== impl HttpBodyService ====
+
+
+struct HttpBodyService<S> {
+    service: S,
+}
+
+impl<S> Service for HttpBodyService<S>
+where
+    S: Service<
+        Request=http::Request<HttpBody>,
+    >,
+{
+    type Request = http::Request<tower_h2::RecvBody>;
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        self.service.poll_ready()
+    }
+
+    fn call(&mut self, req: Self::Request) -> Self::Future {
+        self.service.call(req.map(|b| HttpBody::Http2(b)))
+    }
+}
+
+#[derive(Clone)]
+struct HttpBodyNewService<N> {
+    new_service: N,
+}
+
+impl<N> NewService for HttpBodyNewService<N>
+where
+    N: NewService<Request=http::Request<HttpBody>>,
+    N::Future: 'static,
+{
+    type Request = http::Request<tower_h2::RecvBody>;
+    type Response = N::Response;
+    type Error = N::Error;
+    type Service = HttpBodyService<N::Service>;
+    type InitError = N::InitError;
+    type Future = Box<Future<Item=Self::Service, Error=Self::InitError>>;
+
+    fn new_service(&self) -> Self::Future {
+        Box::new(self.new_service.new_service().map(|s| {
+            HttpBodyService {
+                service: s,
+            }
+        }))
+    }
+}
+
+// ===== impl HyperConnect =====
+
+pub struct HyperConnect<C> {
+    connect: C,
+}
+
+impl<C> hyper::client::Service for HyperConnect<C>
+where
+    C: Connect,
+    C::Future: 'static,
+{
+    type Request = hyper::Uri;
+    type Response = C::Connected;
+    type Error = io::Error;
+    type Future = Box<Future<Item=Self::Response, Error=Self::Error>>;
+
+    fn call(&self, _uri: Self::Request) -> Self::Future {
+        Box::new(self.connect.connect()
+            .map_err(|_e| io::Error::new(io::ErrorKind::Other, "blah")))
+    }
+}
```

Comment:
this file feels large enough that it would benefit from being split into a directory with multiple files?